Text Annotation Frames

claudiaareneee opened the pull request:
## Description

I've been working on the text annotation wishlist items like frames and fillcolor. 

Core changes:
- Created a new folder in the backend for `annotations` and moved related files into it. 
  - `FrameGeometry.ts` (new)
    - I had fun reliving my highschool geometry class for this. But this file is just a bunch of helper methods for calculating the curve primitives for different types of frames.
  - `TextAnnotationElement.ts`
    - Removed `updateAnnotation`.
    - Users should still call `setAnnotation` to change just the annotation of the element. But the geometry is no longer generated here. Instead, geometry is calculated using `TextAnnotationGeometry.appendTextAnnotationGeometry`.
    - Currently out of covenience, I'm calling the `TextAnnotationGeometry.appendTextAnnotationGeometry` in the `toJson` methods of each element, but I could be convinced that this is misleading/wrong and that the user should call this method explicitly.
  - `TextAnnotationGeometry.ts` (new)
    - creates and exposes new methods for producing text annotation geometry by appending the frame geometry and textblock geometry to an ElementGeometry.Builder that's passed in.
    - The debug geomtry logic was moved here.
  - `TextBlockGeometry.ts`(formerly `TextAnnotationGeometry`)
    - removed `produceTextAnnotationGeometry`
    - exposes `produceTextBlockGeometry`
    - moves all the debug logic to `TextAnnotationGeometry`
    - now this file is more concise
  - `TextBlockLayout.ts` (formerly `TextAnnotationLayout`)
    - renamed to be more explicit
    - Fixes a bug 
- `core/backend/src/core-backend.ts`: exposes the new files along with `layoutTextBlock`
- `core/common/src/annotation/TextAnnotation.ts`: Adding properties for frame and fill. These should probably be in the TextStyles, but that's a whole other discussion
- `core/common/src/geometry/ElementGeometry.ts`: adding param for overrides to be reused

Tests
- `core/backend/src/test/annotations/TextAnnotation.test.ts`: tests roundtripping, element encoding, etc.
- `core/backend/src/test/annotations/TextBlock.test.ts`: tests the produceTextBlockGeometry method

Display test app changes:
- `backend/Backend.ts`
- `common/DtaRpcInterface.ts`
- `frontend/TextDecoration.ts`

## Todos / questions

- [x] Comments -> extract API
- [x] Documentation

## Link to Issue

* closes #7677

pmconne commented on core/backend/src/strokers/TextAnnotationStroker.ts:
```diff
@@ -0,0 +1,174 @@
+  }
+
+  public override async computeLayoutResults(props: TextAnnotationStrokerProps, requestProps: StrokerGraphicsRequestProps, options?: TextAnnotationStrokerOptions): Promise<TextAnnotationStrokerResults> {
+    // Todo: Should this be IModelError?
```
It should be `ITwinError`.

claudiaareneee commented on core/common/src/annotation/FrameGeometry.ts:
I had fun reliving my highschool geometry class for this. But this file is just a bunch of helper methods for calculating the curve primitives for different types of frames. It's mostly commented but I need to add a comment about how the computePolygon method doesn't frame as tightly as some of the individual compute methods.

claudiaareneee commented on core/common/src/annotation/TextAnnotation.ts:
Adding properties for frame and fill. These should probably be in the TextStyles, but that's a whole other discussion

claudiaareneee commented on test-apps/display-test-app/src/backend/Backend.ts:
The dta changes are just introducing a new RPC endpoint for the stroker, removing the old endpoint, and adding keyins for the new variables

claudiaareneee commented on core/backend/src/annotations/FrameGeometry.ts:
```diff
@@ -0,0 +1,299 @@
+      return false;
+    }
+
+    // TODO: I need to clean this up. The geom param changes are straddled between this stroker and ElementGeometry.Builder.
+    const params = new GeometryParams(Id64.invalid);
+    params.elmPriority = 0;
+
+    if (frame.fill === undefined) {
+      params.fillDisplay = FillDisplay.Never;
+    } else if (frame.fill === "background") {
+      params.backgroundFill = BackgroundFill.Outline;
+      params.fillDisplay = FillDisplay.Always;
+    } else if (frame.fill !== "subcategory") {
+      params.fillColor = ColorDef.fromJSON(frame.fill);
+      params.lineColor = params.fillColor;
+      params.fillDisplay = FillDisplay.Always;
+    }
+
+    if (frame.border !== "subcategory") {
+      params.lineColor = ColorDef.fromJSON(frame.border);
+      params.weight = frame.borderWeight;
+    }
```
@pmconne, I remember you had feelings about using `FillDisplay.Always` here vs `BlankingFill`, but I can't remember what they are. I'm using `Always` (in combo with `params.elmPriority`) because `Blanking` is applying to all geometry and not just the geometry internal to my builder. I'm up for suggestions. 

Also I'm having an issue with using the `elmPriority` to ensure the text is drawn over the background. Currently I'm having to stradle the `elemPriority` accross `FrameGeometry` and `ElementGeometry` to make sure the params apply propperly. Instead I just want to have it in my `TextAnnotationGeometry` file. I'm still investigating this, but if you have any opinions on this, that would be helpful.

claudiaareneee replied:
Maybe this will help. These are the different configurations I've tried:

https://github.com/user-attachments/assets/a66de461-6160-4ea1-ad10-e300d7592457

claudiaareneee replied:
If not element priority, what should I be using instead?

claudiaareneee replied:
@pmconne, thoughts?

pmconne replied:
Supply .bim file(s) containing the problematic text.

claudiaareneee replied:
Zip: [house.bim.zip](https://github.com/user-attachments/files/20214154/house.bim.zip)

Copy and pastable keyins:
```txt
dta text init
dta text center
dta text height 0.25 // Text size (text height)
dta text width 6 // Width of the text box
dta text color #FDFCDC
dta text text "You were the chosen one! It was said that you would destroy the Sith, not join them. You were to bring balance to the Force, not leave it in darkness."
dta text break
dta text break
dta text text "— Obi-wan Kenobi, "
dta text underline
dta text text "Revenge of the Sith"
dta text margin all 0.25
dta text frame shape capsule
dta text frame border #0081A7
dta text frame borderWeight 10
dta text frame fill #00AFB9"
```

pmconne replied:
Ok so you are not actually putting elements into the iModel. I wanted to see the geometry streams. Do I have to copy-patse each line one at a time into display-test-app's key-in field?

pmconne replied:
nvm `dta macro` exists.

pmconne replied:
The tile generator does not produce edges for shapes with blanking fill. Blanking fill controls order of display within a single element. Element priority controls order of display between different elements. You shouldn't hard-code any specific priority values in your geometry generation code. Your app probably should provide tools to change priorities of 2d elements relative to one another (a la PowerPoint's "move to front/back", "move backward/forward", etc.

I pushed fixes. The frame now draws in front of the fill.

I failed to reproduce your issue with the building geometry displaying in front of the fill. If that persists, please start a separate comment thread with details and repro steps.

pmconne commented:
> @pmconne, thank you for your suggestions, but I still have an issue with blanking fill. Here's a quick video (with audio). I'm using the same bim file from earlier:

To summarize: with latest changes, open Elevation-Front Views - View 1 and run the key-ins. The house geometry appears in front of the text fill. Something different happens in 3d which we will ignore for now.

I can't copy key-ins from a video, can you please provide them?

If you replace `Blanking` with `Always` and change nothing else, this problem vanishes?

claudiaareneee commented:
Just changing blanking fill in `FrameGeometry` L36 & L37 to 'Always' produces this: 
![image](https://github.com/user-attachments/assets/81c97d9f-056c-45cf-bf7e-3a4f5d991e9e)

claudiaareneee commented:
@pmconne, it's possible that this view is just built weird. I tested on a different iModel and it seemed to be working fine

pmconne commented:
> @pmconne, it's possible that this iModel is just built weird. I tested on a different iModel and it seemed to be working fine

Nope, there's a bug with blanking fill. It's been broken forever. I'll fix it in a separate PR.

pmconne commented:
@claudiaareneee for your 3d weirdness, see if toggling Force Surface Discard ON changes things.
![image](https://github.com/user-attachments/assets/070be0ee-c9d5-4c05-a219-905f8a11927b)

claudiaareneee commented:
@pmconne, I'll check and see what happens with that. 

However, at the moment I'm stuck on tests. I can't get it to insert the text element in the `core/backend/src/test/annotations/TextAnnotation.test.ts` persistance tests. I have one suite for `TextAnnotation2d` and another for `TextAnnotation3d`. (The original code only tested the 3d, so I've added another set for 2d). Both test suites are failing at the same place. 

It keeps failing on the `el0.insert()` call.  I believe I've isolated the error to the geometry builder's `appendTextString` method but I could be wrong. The failures are only happening when I have a textblock with content to insert.

✅ <span style="color:green">roundtrips an empty annotation</span>: passes
✅ <span style="color:green">roundtrips an annotation with a style</span>: passes
❌ <span style="color:red">roundtrips an annotation with a textBlock</span>: fails

The error I'm getting is:

```
Error inserting element [BuildGeometryStream failed]
    at Elements.insertElement (/Users/claudianelson/Documents/dp-cospace/repos/itwinjs-core/core/backend/src/IModelDb.ts:782:7)
    at TextAnnotation2d.insert (/Users/claudianelson/Documents/dp-cospace/repos/itwinjs-core/core/backend/lib/cjs/Element.js:183:104)
    at test (/Users/claudianelson/Documents/dp-cospace/repos/itwinjs-core/core/backend/src/test/annotations/TextAnnotation.test.ts:286:22)
    at Context.<anonymous> (/Users/claudianelson/Documents/dp-cospace/repos/itwinjs-core/core/backend/src/test/annotations/TextAnnotation.test.ts:311:75)
    at callFn (/Users/claudianelson/Documents/dp-cospace/repos/itwinjs-core/common/temp/node_modules/.pnpm/mocha@11.1.0/node_modules/mocha/lib/runnable.js:364:21)
    at Test.Runnable.run (/Users/claudianelson/Documents/dp-cospace/repos/itwinjs-core/common/temp/node_modules/.pnpm/mocha@11.1.0/node_modules/mocha/lib/runnable.js:352:5)
    at Runner.runTest (/Users/claudianelson/Documents/dp-cospace/repos/itwinjs-core/common/temp/node_modules/.pnpm/mocha@11.1.0/node_modules/mocha/lib/runner.js:677:10)
    at /Users/claudianelson/Documents/dp-cospace/repos/itwinjs-core/common/temp/node_modules/.pnpm/mocha@11.1.0/node_modules/mocha/lib/runner.js:800:12
    at next (/Users/claudianelson/Documents/dp-cospace/repos/itwinjs-core/common/temp/node_modules/.pnpm/mocha@11.1.0/node_modules/mocha/lib/runner.js:592:14)
    at /Users/claudianelson/Documents/dp-cospace/repos/itwinjs-core/common/temp/node_modules/.pnpm/mocha@11.1.0/node_modules/mocha/lib/runner.js:602:7
    at next (/Users/claudianelson/Documents/dp-cospace/repos/itwinjs-core/common/temp/node_modules/.pnpm/mocha@11.1.0/node_modules/mocha/lib/runner.js:485:14)
    at Immediate._onImmediate (/Users/claudianelson/Documents/dp-cospace/repos/itwinjs-core/common/temp/node_modules/.pnpm/mocha@11.1.0/node_modules/mocha/lib/runner.js:570:5)
    at processImmediate (node:internal/timers:478:21) {
  errorNumber: 65539,
  metadata: {
    elProps: {
      classFullName: 'BisCore:TextAnnotation2d',
      code: [Code],
      model: '0x24',
      jsonProperties: [Object],
      category: '0x16',
      placement: [Placement2d],
      elementGeometryBuilderParams: [Object]
    }
  }
}
        1) roundtrips an annotation with a textBlock
```

I believe the error I'm getting is [IModelStatus.BadArg](https://github.com/iTwin/itwinjs-core/blob/ba8d02fbf9a42c1f6260ff169486c88b03daaae7/common/api/core-bentley.api.md?plain=1#L840) unless I'm missreading that. 

I went ahead and pushed these tests. Looking at lines 311 (test that's failing), 286 (the line that's failing).

As a test, I added a backend ipc method to display-test-app to make sure that I can insert the TextAnnotation2d element. That's working. So something is wrong with the tests. Any thoughts?

I think I must not be mocking something that should be? But I don't know...

pmconne commented:
> Error inserting element [BuildGeometryStream failed]

That's [this function](https://github.com/iTwin/imodel-native/blob/df376a539940576f918bb64dcc2fe923f727fd81/iModelCore/iModelPlatform/DgnCore/ElementGeometry.cpp#L3305). It doesn't like your element geometry builder params.
Show what `elementGeometryBuilderParams` looks like.

> As a test, I added a backend ipc method to display-test-app to make sure that I can insert the TextAnnotation2d element. 

Using the exact same seed iModel that you are creating a copy of in your test?
Check what version of the BisCore schema that seed iModel contains.
Try changing your test to create a brand-new iModel from scratch instead of creating one from a seed (usually preferable for many reasons, including ensuring you have up to date versions of schemas).

claudiaareneee commented:
@pmconne, I did not do my manual test with the same imodel, so let me see what happens there. Do you know of any tests that create an iModel that I could use as an example?

To me the elementGeometryBuilderParams seem normal, but I can't parse them. This is what they look like:

```
Claudia Test >  elementGeometryBuilderParams [
  {
    opcode: 4,
    data: Uint8Array(12) [
      8, 0, 0, 0, 4,
      0, 4, 0, 4, 0,
      0, 0
    ]
  },
  {
    opcode: 22,
    data: Uint8Array(200) [
       20,   0,   0,   0,   0,   0,  14,   0, 116,   0, 115,   0,
        0,   0, 108,   0, 104,   0,   4,   0,  14,   0,   0,   0,
      138, 175, 245, 151, 192, 254, 239,  63,  39, 220, 247, 201,
       88, 222, 145, 191,  39, 101, 234,   3,  29, 246,  51, 191,
      202,   3, 132, 200,  39,  77,  48, 192,  30, 221, 137,  43,
       11, 223, 145,  63, 180, 130,  95,  60, 129, 253, 239,  63,
       39, 220, 247, 201,  88, 222, 145,  63, 143, 120,  92, 236,
       94,  35,  23, 192,   0,   0,   0,   0,   0,   0,   0,   0,
       30, 221, 137,  43,
      ... 100 more items
    ]
  },
  {
    opcode: 22,
    data: Uint8Array(192) [
       20,   0,   0,   0,   0,   0,  14,   0, 116,   0, 115,   0,
        0,   0, 108,   0, 104,   0,   4,   0,  14,   0,   0,   0,
      138, 175, 245, 151, 192, 254, 239,  63,  39, 220, 247, 201,
       88, 222, 145, 191,  39, 101, 234,   3,  29, 246,  51, 191,
      138,  51, 133, 250, 231,  85,  33, 192,  30, 221, 137,  43,
       11, 223, 145,  63, 180, 130,  95,  60, 129, 253, 239,  63,
       39, 220, 247, 201,  88, 222, 145,  63, 128, 144, 148, 137,
      237, 154,  22, 192,   0,   0,   0,   0,   0,   0,   0,   0,
       30, 221, 137,  43,
      ... 92 more items
    ]
  },
  {
    opcode: 22,
    data: Uint8Array(200) [
       20,   0,   0,   0,   0,   0,  14,   0, 116,   0, 115,   0,
        0,   0, 108,   0, 104,   0,   4,   0,  14,   0,   0,   0,
      138, 175, 245, 151, 192, 254, 239,  63,  39, 220, 247, 201,
       88, 222, 145, 191,  39, 101, 234,   3,  29, 246,  51, 191,
      182, 173,  46, 151,  24,  49,  16, 192,  30, 221, 137,  43,
       11, 223, 145,  63, 180, 130,  95,  60, 129, 253, 239,  63,
       39, 220, 247, 201,  88, 222, 145,  63, 107, 101, 209,  37,
       39, 149,  19, 192,   0,   0,   0,   0,   0,   0,   0,   0,
       30, 221, 137,  43,
      ... 100 more items
    ]
  },
  {
    opcode: 10,
    data: Uint8Array(88) [
       98, 103,  48,  48,  48,  49, 102,  98,  12,   0,   0,   0,
        8,   0,  10,   0,   9,   0,   4,   0,   8,   0,   0,   0,
       12,   0,   0,   0,   0,   1,   6,   0,   4,   0,   4,   0,
        6,   0,   0,   0, 240, 187, 113,  20, 248,  45,  16, 192,
       19,  73, 240,  95,  76,  72,  20, 192,   0,   0,   0,   0,
        0,   0,  89,  64,   0,   0,   0,   0,   0, 240, 226,  60,
      255, 255, 255, 255, 255, 255,  19, 192,   0,   0,   0,   0,
        0,   0,  89,  64
    ]
  },
  {
    opcode: 22,
    data: Uint8Array(192) [
       20,   0,   0,   0,   0,   0,  14,   0, 116,   0, 115,   0,
        0,   0, 108,   0, 104,   0,   4,   0,  14,   0,   0,   0,
      138, 175, 245, 151, 192, 254, 239,  63,  39, 220, 247, 201,
       88, 222, 145, 191,  39, 101, 234,   3,  29, 246,  51, 191,
       33, 213,  42,  75,  91,  31,  10, 192,  30, 221, 137,  43,
       11, 223, 145,  63, 180, 130,  95,  60, 129, 253, 239,  63,
       39, 220, 247, 201,  88, 222, 145,  63, 142, 209,  24,  55,
       93, 186,  23, 192,   0,   0,   0,   0,   0,   0,   0,   0,
       30, 221, 137,  43,
      ... 92 more items
    ]
  }
]
```

pmconne commented:
> Do you know of any tests that create an iModel that I could use as an example?

See the tests added in #7785 (you can steal parts of the backend code).

claudiaareneee commented:
Cool cool, I'll do that. It probably is a schema issue. Running the manual test I got a "unable to enter graphical editing scope: version too old" error. I'm not sure of the actual schema number yet though

claudiaareneee commented:
Heyyo, my tests work now so I'm publishing this PR

claudiaareneee commented on core/backend/src/annotations/TextAnnotationElement.ts:
```diff
@@ -63,8 +57,8 @@ export class TextAnnotation2d extends AnnotationElement2d {
    * @param subCategory If specified, the subcategory on which to define the geometry; otherwise, the default subcategory of the element's category is used.
```
TODO: fix these comments, I forgot to remove subcategory.

pmconne commented:
> Heyyo, my tests work now so I'm publishing this PR

Can you make sure the PR description is up to date? I see a lot of unchecked TODOs.

claudiaareneee commented:
@pmconne, I found this discussion: https://github.com/iTwin/itwinjs-core/issues/6793#issuecomment-2150832935.

Essentially because of the refactoring of the text annotation geometry, I now need `TextBlockLayout` in more places, and the `extract-api` command was failing. I ended up making some of the classes in this file `@beta` for this reason. 

If you think this API needs to be cleaned up, could I do that in a separate PR? This one is already really large and I don't want to oversaturate it.

pmconne commented on core/backend/src/annotations/FrameGeometry.ts:
```diff
@@ -0,0 +1,324 @@
+   * @returns `true` if any geometry was appended to the builder
+  * @beta
+   */
+  export const appendFrameToBuilder = (builder: ElementGeometry.Builder, frame: TextFrameStyleProps, range: Range2d, transform: Transform, geomParams?: GeometryParams): boolean => {
```
@ben-polinsky will documentation generator produce correct docs for functions declared like this instead of e.g. `export function(whatever)`?
@claudiaareneee was this intended to prevent weirdos from doing `FrameGeometry.appendFrameToBuilder = () => throw new Error("mwuhaha");`?
(IMO the namespace is just API clutter and you could remove it).

claudiaareneee replied:
Since it's broken, I change it to export them as functions.

claudiaareneee replied:
Done

pmconne commented on core/backend/src/annotations/FrameGeometry.ts:
```diff
@@ -0,0 +1,324 @@
+      return false;
+    }
+
+    const params = geomParams?.clone() ?? new GeometryParams(Id64.invalid);
```
Why `Id64.invalid`? Does anyone actually pass in `undefined` for `geomParams`?

claudiaareneee replied:
I can't remember where I got that from. Should I be using the element's category id?

claudiaareneee replied:
Okay I remember where I got that from. The ElementGeometry.builder and GeometryStreamBuilder both use Id64.Invalid for the lines in the fractions bars.

Maybe the frame geom should be a different category though. Thoughts @pmconne?

pmconne replied:
That's for your app and/or the user to decide. Either way, I think your API here should require the caller to configure all of that - you certainly can't decide here what category to put the frame on. If you want to continue just cloning the input GeometryParams, I think you should make them non-optional so you can eliminate Id64.invalid here.

claudiaareneee replied:
Done! 

I think that should be the last comment to address!

pmconne commented on core/backend/src/annotations/FrameGeometry.ts:
```diff
@@ -0,0 +1,324 @@
+
+  /**
+   * Computes the frame geometry based on the provided frame shape and range.
+   * @returns a [[Loop]] or [[Path]](if it's just a line) that represents the frame geometry
```
```suggestion
   * @returns a [Loop]($geometry) or [Path]($geometry) (if it's just a line) that represents the frame geometry
```
`[[SomeThing]]` is short-hand for `[SomeThing]($package-containing-this-file)`. It doesn't work for types defined outside of this package.
@ben-polinsky why isn't that producing a documentation build error about `Loop` and `Path` not being defined in core-backend?

ben-polinsky replied:
I'm not sure, [I'll have to investigate](https://github.com/iTwin/itwinjs-core/issues/8159).

claudiaareneee replied:
> `[[SomeThing]]` is short-hand for `[SomeThing]($package-containing-this-file)`

Good to know! I did not know that

pmconne commented on core/backend/src/annotations/FrameGeometry.ts:
```diff
@@ -0,0 +1,324 @@
+  }
+
+  /** Line - currently just adds an underline. Once we have leaders, this method may change. */
+  const computeLine = (range: Range2d, transform: Transform): Path => {
```
If you're going to use a namespace, move all the non-exported stuff outside of it so it's clear at a glance what's being exported vs local implementation detail.

pmconne commented on core/common/src/annotation/TextAnnotation.ts:
```diff
@@ -35,6 +36,31 @@ export interface TextAnnotationAnchor {
+ * @beta
+ */
+export interface TextFrameStyleProps {
+  shape: TextAnnotationFrameShape;
```
Did you consider defaulting to "rectangle"?

pmconne commented on core/common/src/annotation/TextAnnotation.ts:
```diff
@@ -115,6 +148,7 @@ export class TextAnnotation {
       orientation: props?.orientation ? YawPitchRollAngles.fromJSON(props.orientation) : undefined,
       textBlock: props?.textBlock ? TextBlock.create(props.textBlock) : undefined,
       anchor: props?.anchor ? { ...props.anchor } : undefined,
+      frame: props?.frame,
```
You should probably clone this to prevent unwanted unintentional modifications via the input object.

claudiaareneee replied:
done

pmconne commented on core/common/src/annotation/TextAnnotation.ts:
```diff
@@ -140,6 +174,8 @@ export class TextAnnotation {
       props.anchor = { ...this.anchor };
     }
 
+    props.frame = this.frame;
```
You should definitely clone this so that the caller can modify the returned JSON object without accidentally mutating the source object.

pmconne reviewed (approved):
Comments are mostly nitpickery.

claudiaareneee commented on core/backend/src/annotations/TextBlockLayout.ts:
```diff
@@ -84,7 +85,7 @@ export function layoutTextBlock(args: LayoutTextBlockArgs): TextBlockLayout {
  * Gets the result of laying out the the contents of a TextBlock into a series of lines containing runs.
  * The visual layout accounts for the [TextStyle]($common)s, fonts, and [TextBlock.width]($common). It applies word-wrapping if needed.
- * The layout returned matches the visual layout of the geometry produced by [[produceTextAnnotationGeometry]].
+ * The layout returned matches the visual layout of the geometry produced by [[TextAnnotationGeometry]].
```
Would you rather I un-`namespace` everything? I thought that looked cleaner to me, but I'm not attached to it. If you think that it would be better to not wrap things in a `namespace`, I can get rid of them

pmconne replied:
Namespaces are another thing that mostly help the writer of the code (they don't have to type the namespace inside the namespace) and not the callers/readers (they have to type the namespace all over the place).

pmconne commented on core/common/src/annotation/TextAnnotation.ts:
```diff
@@ -35,6 +36,31 @@ export interface TextAnnotationAnchor {
+/** Set of predefined shapes that can be computed and drawn around the margins of a [[TextBlock]]
+ * @beta
+*/
+export type TextAnnotationFrameShape = "none" | "line" | "rectangle" | "circle" | "equilateralTriangle" | "diamond" | "square" | "pentagon" | "hexagon" | "octagon" | "capsule" | "roundedRectangle";
```
> I would think we would only add more and not take away. I don't think the change would be very breaking.

Unfortunately tsc is strict about exhaustiveness, so adding a new enum member can cause existing code to fail to compile. That's a general problem that we don't have a solution for yet. Since these APIs are currently `@beta` let's leave it as-is for now.

claudiaareneee commented on core/common/src/annotation/TextAnnotation.ts:
```diff
@@ -35,6 +36,31 @@ export interface TextAnnotationAnchor {
+  shape: TextAnnotationFrameShape;
+  fill?: TextAnnotationFillColor;
+  border?: TextStyleColor;
+  borderWeight?: number;
```
I considered that, but I found this approach to be similar.  

> Weight is meaningful only if border is defined. Did you consider

It's more like:
Weight is meaningful only if **shape** is defined.

There is a default color (black) so the user can still define borderWeight meaningfully.

claudiaareneee replied:
done! sorry, I was in the process of committing those changes when I sent that last message

claudiaareneee commented on core/backend/src/annotations/TextAnnotationGeometry.ts:
```diff
@@ -0,0 +1,110 @@
+ * @beta
+ * @see [[appendTextAnnotationGeometry]] to append the geometry to an [[ElementGeometry.Builder]].
+ */
+export interface RequestProps {
```
My bad, was left over from the namespace. Changed!
